relay-pattern/src/wildmatch.rs

```diff
@@ -0,0 +1,605 @@
+        true => is_match_impl::<_, CaseInsensitive>(tokens.as_slice(), haystack),
+    }
+}
+
+/// Bundles necessary matchers for [`is_match_impl`].
+trait Matcher {
+    /// Returns the length of the `needle` in the `haystack` if the `needle` is a prefix of `haystack`.
+    fn is_prefix(haystack: &str, needle: &Literal) -> Option<usize>;
+    /// Searches for the `needle` in the `haystack` and returns the index of the start of the match
+    /// and the length of match or `None` if the `needle` is not contained in the `haystack`.
+    fn find(haystack: &str, needle: &Literal) -> Option<(usize, usize)>;
+    /// Returns `true` if the char `c` is contained within `ranges`.
+    fn ranges_match(c: char, negated: bool, ranges: &Ranges) -> bool;
+    /// Searches for the first char in the `haystack` that is contained in one of the `ranges`.
+    ///
+    /// Returns the offset in bytes and matching `char` if the range is contained within the
+    /// `haystack`.
+    #[inline(always)]
+    fn ranges_find(haystack: &str, negated: bool, ranges: &Ranges) -> Option<(usize, char)> {
+        // TODO: possibly optimize range finding.
+        // TODO: possibly optimize with `memchr{1,2,3}` for short ranges.
+        haystack
+            .char_indices()
+            .find(|&(_, c)| Self::ranges_match(c, negated, ranges))
+    }
+}
+
+/// A case sensitive [`Matcher`].
+struct CaseSensitive;
+
+impl Matcher for CaseSensitive {
+    #[inline(always)]
+    fn is_prefix(haystack: &str, needle: &Literal) -> Option<usize> {
+        let needle = needle.as_case_converted_bytes();
+        memchr::arch::all::is_prefix(haystack.as_bytes(), needle).then_some(needle.len())
+    }
+
+    #[inline(always)]
+    fn find(haystack: &str, needle: &Literal) -> Option<(usize, usize)> {
+        let needle = needle.as_case_converted_bytes();
+        memchr::memmem::find(haystack.as_bytes(), needle).map(|offset| (offset, needle.len()))
+    }
+
+    #[inline(always)]
+    fn ranges_match(c: char, negated: bool, ranges: &Ranges) -> bool {
+        ranges.contains(c) ^ negated
+    }
+}
+
+/// A case insensitive [`Matcher`].
+struct CaseInsensitive;
+
+impl Matcher for CaseInsensitive {
+    #[inline(always)]
+    fn is_prefix(haystack: &str, needle: &Literal) -> Option<usize> {
+        // We can safely assume `needle` is already full lowercase. This transformation is done on
+        // token creation based on the options.
+        //
+        // The haystack cannot be converted to full lowercase to not break class matches on
+        // uppercase unicode characters which would produce multiple lowercase characters.
+        //
+        // TODO: benchmark if allocation free is better/faster.
+        let needle = needle.as_case_converted_bytes();
+        let lower_haystack = haystack.to_lowercase();
+
+        memchr::arch::all::is_prefix(lower_haystack.as_bytes(), needle)
+            .then(|| recover_offset_len(haystack, 0, needle.len()).1)
```

Comment:
Do we have enough tests that cover unicode strings with both `is_prefix` and `find`?

Reply:
I added a bunch more tests for both `Matcher` implementations.